Sources/VariadicsGenerator/VariadicsGenerator.swift

```diff
@@ -0,0 +1,186 @@
+// swift run VariadicsGenerator --max-arity 10 > Sources/RegexDSL/Concatenation.swift
+
+import ArgumentParser
+
+struct Permutation {
+  let arity: Int
+  // 1 -> no extra constraint
+  // 0 -> where T.CaptureValue: NoCaptureProtocol
+  let bits: Int64
+
+  func isCaptureless(at index: Int) -> Bool {
+    bits & (1 << index) != 0
+  }
+
+  var hasCaptureless: Bool {
+    bits != 0
+  }
+
+  func hasCaptureless(beyond index: Int) -> Bool {
+    bits >> (index + 1) != 0
+  }
+
+  var identifier: String {
+    var result = ""
+    for i in 0..<arity {
+      result.append(isCaptureless(at: i) ? "V" : "T")
+    }
+    return String(result.reversed())
+  }
+}
+
+struct Permutations: Sequence {
+  let arity: Int
+
+  struct Iterator: IteratorProtocol {
+    let arity: Int
+    var counter = Int64(0)
+
+    mutating func next() -> Permutation? {
+      guard counter & (1 << arity) == 0 else {
+        return nil
+      }
+      defer { counter += 1 }
+      return Permutation(arity: arity, bits: counter)
+    }
+  }
+
+  public func makeIterator() -> Iterator {
+    Iterator(arity: arity)
+  }
+}
+
+func output(_ content: String) {
+  print(content, terminator: "")
+}
+
+typealias Counter = Int64
+let patternProtocolName = "RegexProtocol"
+let concatenationStructTypeBaseName = "Concatenate"
+let matchValueAssociatedTypeName = "MatchValue"
+let captureValueAssociatedTypeName = "CaptureValue"
+let patternBuilderTypeName = "RegexBuilder"
+let patternProtocolRequirementName = "regex"
+let PatternTypeBaseName = "Regex"
+let emptyProtocolName = "EmptyProtocol"
+let emptyStructName = "Empty"
+
+@main
+struct VariadicsGenerator: ParsableCommand {
+  @Option(help: "The minimum arity of declarations to generate.")
+  var minArity: Int = 2
+
+  @Option(help: "The maximum arity of declarations to generate.")
+  var maxArity: Int
+
+  func run() throws {
+    precondition(minArity > 0)
+    precondition(maxArity > 1)
+    precondition(maxArity < Counter.bitWidth)
+
+    output("// BEGIN AUTO-GENERATED CONTENT\n\n\n")
+
+    for arity in minArity...maxArity {
+      emitDeclarations(arity: arity)
+      output("\n\n")
+    }
+
+    output("// END AUTO-GENERATED CONTENT")
+  }
+
+  func emitDeclarations(arity: Int) {
+    for permutation in Permutations(arity: arity) {
+      // Emit type declarations.
+      let typeName = "\(concatenationStructTypeBaseName)\(arity)_\(permutation.identifier)"
+      output("public struct \(typeName)<")
+      for i in 0..<arity {
+        output("T\(i): \(patternProtocolName)")
+        if i != arity - 1 {
+          output(", ")
+        }
```

Comment:
I would definitely extract this out into a helper method. I know this is a throw-away script, but it should make the function easier to read and I (side project) want to keep improving string formatting.

Reply:
Done!